PCF-ADR-0007-operator-redesign.md

```diff
@@ -0,0 +1,201 @@
+
+### Configuration
+
+Each controller, i.e., MCAD and InstaScale, as well as the CodeFlare operator, must provide a unified configuration.
+
+Currently each component has its own way of sourcing configuration:
+* CLI options for MCAD
+* CLI options, and a ConfigMap for InstaScale
+* CLI options, and CRDs for the CodeFlare operator
+
+We propose to model the configuration of each component as part of their public API.
+
+MCAD would declare its configuration, e.g.:
+
+```go
+package config
+
+struct MCADConfiguration {
+	// Number of seconds a job will go away for, if it can not be scheduled.
+	// Default is 20.
+	BackoffTime int `json:"backoffTime,omitempty"`
+	// Head of line job will not be bumped away for at least HeadOfLineHoldingTime seconds
+	// by higher priority jobs.
+	// Default setting to 0 disables this mechanism.
+	HeadOfLineHoldingTime int `json:"headOfLineHoldingTime,omitempty"`
+
+	// Remaining configuration...
+}
+```
+
+InstaScale would also declare its configuration, e.g.:
+
+```go
+package config
+
+struct InstaScaleConfiguration {
+	// Maximum of cluster nodes that can be scaled out.
+	// Defaults to 3.
+	MaxScaleoutAllowed int `json:"maxScaleoutAllowed,omitempty"`
+
+	// Remaining configuration...
+}
+```
+
+Finally, the CodeFlare operator would embed MCAD and InstaScale configurations, along with the generic configuration taken from [component-base](https://github.com/kubernetes/component-base), e.g.:
+
+```go
+package config
+
+import (
+	configv1alpha1 "k8s.io/component-base/config/v1alpha1"
+	mcad "github.com/project-codeflare/multi-cluster-app-dispatcher/pkg/config"
+	instascale "github.com/project-codeflare/instascale/pkg/config"
+)
+
+struct InstaScaleConfiguration {
+	// Enabled controls whether the InstaScale controller is started.
+	// It may defaults to true on platforms that InstaScale supports.
+	// Otherwise defaults to false.
+	Enabled *bool `json:"enabled,omitempty"`
+
+	// The InstaScale controller configuration
+	instascale.InstaScaleConfiguration `json:",inline,omitempty"`
+}
+
+struct CodeFlareOperatorConfiguration {
+	// The MCAD controller configuration
+	MCAD *mcad.MCADConfiguration `json:"mcad,omitempty"`
+
+	// The InstaScale controller configuration
+	InstaScale *InstaScaleConfiguration `json:"instascale,omitempty"`
+
+	// ClientConnection configures the connection to the cluster
+	ClientConnection *configv1alpha1.ClientConnectionConfiguration `json:"clientConnection,omitempty"`
+
+	// LeaderElection is the LeaderElection config to be used when configuring
+	// the manager.Manager leader election
+	LeaderElection *configv1alpha1.LeaderElectionConfiguration `json:"leaderElection,omitempty"`
+
+	// Remaining configuration...
+}
+```
+
+The later `struct` would be sourced from a ConfigMap, mounted into the operator deployment container.
```

Comment:
I wonder if there's a better way to do this, like if a user wants to update the configuration _after_ the operator is up and running.

Reply:
A mounted configMap is updated automatically with new values, configmap values used as env variables do not however. Making sure the configuration is kept up to date by re-reading the files in the mounted directory would be the job of the operator IMO.

Reply:
While the ConfigMap will automatically be updated at it's mounted as a volume, it won't be hot-reloaded.

As some of the keys configure the controller-runtime manager, this requires a restart anyway, which is best achieved by restarting the Pod. It might be possible to implement hot-reloading, but the complexity / cost would overweight the value.

Usually, restarting Deployment Pod(s) upon mounted ConfigMap changes, is achieved by external controllers / operators. There are lots of implementations available already, which have emerged as a solution to kubernetes/kubernetes#22368.